src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,527 @@
+    // Notify the remote output service that the build is about to
+    // start. The remote output service will return the directory in
+    // which it wants us to let the build take place.
+    //
+    // Make the output service aware of the location of the output path
+    // from our perspective and any symbolic links that will point to
+    // it. This allows the service to properly resolve symbolic links
+    // containing absolute paths as part of BatchStat() calls.
+    StartBuildRequest.Builder builder = StartBuildRequest.newBuilder();
+    builder.setOutputBaseId(outputBaseId);
+    builder.setBuildId(buildId.toString());
+    builder.setInstanceName(instanceName);
+    builder.setDigestFunction(digestFunction);
+    builder.setOutputPathPrefix(outputPathPrefix.toString());
+    Path originalOutputPath = execRoot.getRelative(relativeOutputPath);
+    builder.putOutputPathAliases(originalOutputPath.toString(), ".");
+    StartBuildResponse response = blockingStub.startBuild(builder.build());
+
+    // Replace the output path with a symbolic link pointing to the
+    // directory managed by the remote output service.
+    PathFragment outputPath = outputPathPrefix.getRelative(response.getOutputPathSuffix());
+    try {
+      try {
+        originalOutputPath.deleteTree();
+      } catch (FileNotFoundException e) {}
+      originalOutputPath.createSymbolicLink(outputPath);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    stateAcrossBuilds.currentBuildId = buildId;
+    currentOriginalOutputPath = originalOutputPath.asFragment();
+    currentRelativeOutputPath = PathFragment.create(relativeOutputPath);
+
+    if (stateAcrossBuilds.previousBuildId == null || !response.hasInitialOutputPathContents()) {
+      // Either Bazel or the remote output service has performed no
+      // build before.
+      return ModifiedFileSet.EVERYTHING_MODIFIED;
+    }
+    InitialOutputPathContents initialContents = response.getInitialOutputPathContents();
+    if (initialContents.getBuildId() != stateAcrossBuilds.previousBuildId.toString()) {
+      // Bazel and the remote output service disagree on the build ID of
+      // the previous build.
+      return ModifiedFileSet.EVERYTHING_MODIFIED;
+    }
+    // Bazel and the remote output service agree on the build ID of the
+    // previous build. Return the set of paths that have been modified.
+    //
+    // TODO: Do these paths need to be relative to the exec root or the
+    // output path? Assume the exec root for now.
+    return ModifiedFileSet.builder()
+        .modifyAll(
+            Iterables.transform(
+                initialContents.getModifiedPathsList(),
+                (p) -> currentRelativeOutputPath.getRelative(p)))
+        .build();
+  }
+
+  @Override
+  public void finalizeBuild(boolean buildSuccessful)
+      throws BuildFailedException, AbruptExitException, InterruptedException {
```

Comment:
nit: delete

Reply:
I think this is actually necessary, right? The `buildSuccessful` value we capture gets passed to the `FinalizeBuild` RPC.